sound/soc/sof/ipc4-telemetry.c

```diff
@@ -0,0 +1,109 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2018-2023 Intel Corporation. All rights reserved.
+//
+
+#include <linux/debugfs.h>
+#include <linux/io.h>
+#include <linux/pm_runtime.h>
+#include <sound/sof/debug.h>
+#include <sound/sof/ipc4/header.h>
+#include "sof-priv.h"
+#include "ops.h"
+#include "ipc4-telemetry.h"
+#include "ipc4-priv.h"
+
+static void __iomem *sof_ipc4_query_exception_address(struct snd_sof_dev *sdev)
+{
+	u32 telemetry_slot_offset;
+	void __iomem *base_addr;
+	u32 offset;
+
+	telemetry_slot_offset =
+		sof_ipc4_find_debug_slot_offset_by_type(sdev,
+							SOF_IPC4_DEBUG_SLOT_TELEMETRY);
+	if (!telemetry_slot_offset)
+		return NULL;
+
+	/* skip the first separator code */
+	offset = telemetry_slot_offset + sizeof(u32);
+	base_addr = sdev->bar[sdev->mailbox_bar] + offset;
+
+	return base_addr;
+}
+
+static ssize_t sof_telemetry_entry_read(struct file *file, char __user *buffer,
+					size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	loff_t pos = *ppos;
+	size_t size_ret;
+
+	if (pos < 0)
+		return -EINVAL;
+	if (pos >= dfse->size || !count)
+		return 0;
+	if (count > dfse->size - pos)
+		count = dfse->size - pos;
+
+	if (!dfse->io_mem)
+		return 0;
+
+	size_ret = copy_to_user(buffer, dfse->cache_buf + pos, count);
+
+	/* update count & position if copy succeeded */
+	if (size_ret)
+		return -EFAULT;
+
+	*ppos = pos + count;
+
+	return count;
+}
+
+static const struct file_operations sof_telemetry_fops = {
+	.open = simple_open,
+	.read = sof_telemetry_entry_read,
+};
+
+void sof_ipc4_create_exception_debugfs_node(struct snd_sof_dev *sdev)
+{
+	struct snd_sof_dfsentry *dfse;
+
+	dfse = devm_kzalloc(sdev->dev, sizeof(*dfse), GFP_KERNEL);
+	if (!dfse)
+		return;
+
+	dfse->name = "exception";
+	dfse->type = SOF_DFSENTRY_TYPE_BUF;
+	dfse->size = SOF_IPC4_DEBUG_SLOT_SIZE - sizeof(u32);
+	dfse->sdev = sdev;
+
+	dfse->cache_buf = devm_kzalloc(sdev->dev, dfse->size, GFP_KERNEL);
+	if (!dfse->cache_buf)
+		return;
+
+	list_add(&dfse->list, &sdev->dfsentry_list);
+
+	debugfs_create_file("exception", 0444, sdev->debugfs_root, dfse, &sof_telemetry_fops);
+}
+
+void sof_ipc4_update_exception_debugfs_node(struct snd_sof_dev *sdev)
```

Comment:
If you rename this function to `sof_ipc4_handle_exception()` and wrap in the call to `snd_sof_handle_fw_exception() here, then things will be simpler?